core/crypto/kem/hybrid/hybrid.go

```diff
@@ -0,0 +1,329 @@
+		return nil, err
+	}
+	return append(first, second...), nil
+}
+
+func (sk *PrivateKey) Equal(other kem.PrivateKey) bool {
+	oth, ok := other.(*PrivateKey)
+	if !ok {
+		return false
+	}
+	if sk.first == nil && sk.second == nil && oth.first == nil && oth.second == nil {
+		return true
+	}
+	if sk.first == nil || sk.second == nil || oth.first == nil || oth.second == nil {
+		return false
+	}
+	return sk.first.Equal(oth.first) && sk.second.Equal(oth.second)
+}
+
+func (sk *PrivateKey) Public() kem.PublicKey {
+	return &PublicKey{sk.scheme, sk.first.Public(), sk.second.Public()}
+}
+
+func (pk *PublicKey) Equal(other kem.PublicKey) bool {
+	oth, ok := other.(*PublicKey)
+	if !ok {
+		return false
+	}
+	if pk.first == nil && pk.second == nil && oth.first == nil && oth.second == nil {
+		return true
+	}
+	if pk.first == nil || pk.second == nil || oth.first == nil || oth.second == nil {
+		return false
+	}
+	return pk.first.Equal(oth.first) && pk.second.Equal(oth.second)
+}
+
+func (pk *PublicKey) MarshalBinary() ([]byte, error) {
+	if pk.first == nil || pk.second == nil {
+		return nil, ErrUninitialized
+	}
+	first, err := pk.first.MarshalBinary()
+	if err != nil {
+		return nil, err
+	}
+	second, err := pk.second.MarshalBinary()
+	if err != nil {
+		return nil, err
+	}
+	return append(first, second...), nil
+}
+
+func (sch *Scheme) GenerateKeyPair() (kem.PublicKey, kem.PrivateKey, error) {
+	pk1, sk1, err := sch.first.GenerateKeyPair()
+	if err != nil {
+		return nil, nil, err
+	}
+	pk2, sk2, err := sch.second.GenerateKeyPair()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return &PublicKey{sch, pk1, pk2}, &PrivateKey{sch, sk1, sk2}, nil
+}
+
+func (sch *Scheme) DeriveKeyPair(seed []byte) (kem.PublicKey, kem.PrivateKey) {
+	if len(seed) != sch.SeedSize() {
+		panic(kem.ErrSeedSize)
+	}
+
+	h, err := blake2b.NewXOF(uint32(sch.first.SeedSize()+sch.second.SeedSize()), nil)
+	if err != nil {
+		panic(err)
+	}
+	seedHash := blake2b.Sum256(seed)
+	_, _ = h.Write(seedHash[:])
+	first := make([]byte, sch.first.SeedSize())
+	second := make([]byte, sch.second.SeedSize())
+	_, _ = h.Read(first)
+	_, _ = h.Read(second)
+
+	pk1, sk1 := sch.first.DeriveKeyPair(first)
+	pk2, sk2 := sch.second.DeriveKeyPair(second)
+
+	return &PublicKey{sch, pk1, pk2}, &PrivateKey{sch, sk1, sk2}
+}
+
+func splitPRF(ss1, ss2, ct1, ct2 []byte) []byte {
+	// implement split PRF KEM combiner as:
+	// H(ss1 || ss2 || ct1 || ct2)
+	// in order to retain IND-CCA2 security
+	// as described here:
+	// https://link.springer.com/chapter/10.1007/978-3-319-76578-5_7
+
+	h, err := blake2b.New256(nil)
+	if err != nil {
+		panic(err)
+	}
+	_, err = h.Write(ss1)
+	if err != nil {
+		panic(err)
+	}
+	_, err = h.Write(ss2)
+	if err != nil {
+		panic(err)
+	}
+	_, err = h.Write(ct1)
+	if err != nil {
+		panic(err)
+	}
+	_, err = h.Write(ct2)
+	if err != nil {
+		panic(err)
+	}
+	return h.Sum(nil)
+}
+
+func (sch *Scheme) Encapsulate(pk kem.PublicKey) (ct, ss []byte, err error) {
```

Comment:
This implementation can just be a thin wrapper around `EncapsulateDeterministically` like in other implementations, then we don't risk the implementations diverging.

Reply:
agreed. fixed.